### PR TITLE
Add torchvision support for the HPatches dataset 

### PR DESCRIPTION
**Pull request by vbalnt**
Hi, 

we wrote a `torchvision` loader for the [HPatches dataset ](https://hpatches.github.io/), which is suitable for training several metric learning methods using contrastive loss, triplet loss, and set based losses. 

Submitting a pull request in case you want to merge it. 

The loader downloads and caches all the files that are needed from the dataset, and it can be used in 3 ways:

### Pairs

Returns pairs of patches, together with corresponding labels. Suitable for siamese networks training with contrastive losses. 

The labels are 0:negative pair (non-matching) and  1:positive pair (matching). 

l
r

    labels : [1, 1, 1, 1, 0, 1, 0, 0, 1, 0]

### Triplets

Returns triplets suitable for training metric learning triplet networks. 
From top to bottom `anchors`, `positives`, `negatives`. 
a
p
n

### Sets

Returns two sets of patches, a `pos` set which contains patches coming from the same point, 
and the `neg` class in which the patches come from different points. This is suitable for generic set-based metric learning.

(`pos`) set of patches
pos
(`neg`) set of patches
neg

### Sample usage 

```
from torchvision import datasets, transforms, utils
t = transforms.Compose([transforms.ToTensor()])

# Pairs
db = datasets.HPatches('./', split='full',  transform=t, download=True, n_samples = 1e7, output='pairs')
for i in range(3):
    sample = db[i]
    print(sample.keys())

# Triplets
db = datasets.HPatches('./', split='full',  transform=t,  n_samples = 1e7, output='triplets')
for i in range(3):
     sample = db[i]
     print(sample.keys())

# Sets
db = datasets.HPatches('./', split='full',  transform=t, n_samples = 1e7, output='sets')
for i in range(3):
     sample = db[i]
     print(sample.keys())
```

### CODE REVIEW

File: torchvision/datasets/hpatches.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            index (int): Index
+        Returns:
+            dict: {'left':, 'right':, 'label':} when `output` is `pairs`.
+                           OR
+            dict: {'anchor':, 'pos':, 'neg':} when `output` is `triplets`.
+                           OR
+            dict: {'pos':, 'neg':} when `output` is `sets`.
+        """
+        # fix the random seed issue with numpy and multiprocessing
+        seed = random.randrange(4294967295)
```
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: @alykhantejani I actually did that due to a weird behaviour of `np.random` which makes all the workers produce the same random numbers, so each mini-batch ended up being identical when run in parallel. 

More here:
https://discuss.pytorch.org/t/why-does-numpy-random-rand-produce-the-same-values-in-different-cores/12005

**Comment by soumith** [comment]
@vbalnt wonder if you had time to look at the comments

**Comment by vbalnt** [comment]
@soumith unfortunately been busy with other stuff, will try to fix them over the weekend.

File: torchvision/datasets/hpatches.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            patches = torch.load(os.path.join(self.root, self.folder, seq, tp + ".pth"))
+            all_patches[tp] = patches
+        d[seq] = all_patches
+
+    def do_check_exists(self):
+        return os.path.exists(os.path.join(self.root, self.folder))
+
+    def do_download(self):
+        p = Pool()
+        from six.moves import urllib
```
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: This seems to already be included in pytorch 

https://github.com/pytorch/vision/blob/47214f0506b8cfa5451574e011e56b043d150fce/torchvision/datasets/utils.py#L30

File: torchvision/datasets/hpatches.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+class HPatches(data.Dataset):
+    """`HPatches <https://hpatches.github.io/>`_ Dataset.
+    Args:
+        root (string): Root directory of dataset where ``hpatches-release``
+            will be downloaded.
+        split (string, optional): If given, only sequences from a specific split will be
+            used. Default is the 'full' split, i.e. all sequences.
+        transform (callable, optional): A function/transform that takes in an PIL image
+            and returns a transformed version. E.g, ``transforms.RandomCrop``.
+        download (bool, optional): If true, downloads the dataset from the internet and
```
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Done, thanks

File: torchvision/datasets/hpatches.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            puts it in root directory. If the dataset is already downloaded, it is not
+            downloaded again.
+        output (string, optional): Type of output from the dataset generator: {pairs,
+            triplets, sets}. Pairs are suitable for siamese networks, triplets for
+            triplet networks, and sets for generic metric learning method. Default is
+            'pairs'.
+        n_samples (int, optional): Number of items to generate at each epoch. Default
+            is 1e7.
+        cache_to_ram (bool, optional): Whether to cache to all sequences to RAM to avoid
```
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Not, sure I am guessing 10gigs of ram are not that bad. 
Perhaps this can be dynamically set? I.e if there is more than 12 or so free ram cache, otherwise not

File: torchvision/datasets/hpatches.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+tps = ['ref', 'e1', 'e2', 'e3', 'e4', 'e5', 'h1', 'h2', 'h3', 'h4', 'h5',
+       't1', 't2', 't3', 't4', 't5']
+
+
+class HPatches(data.Dataset):
+    """`HPatches <https://hpatches.github.io/>`_ Dataset.
+    Args:
+        root (string): Root directory of dataset where ``hpatches-release``
+            will be downloaded.
+        split (string, optional): If given, only sequences from a specific split will be
```
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Added the other splits too

File: torchvision/datasets/hpatches.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: OK, added automatically setting this to `False` if there is not enough available ram.

[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Changed

File: torchvision/datasets/hpatches.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/datasets/hpatches.py]: I believe this can be fixed now in master if you pass the `worker_init_fn` to `DataLoader`, and in there you specify something like
```python
def worker_init_fn(worker):
    np.random.seed(worker + 1234)
```
so can you please remove this?

File: torchvision/datasets/hpatches.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/datasets/hpatches.py]: I think this is fine here.

File: test/test_transforms.py [hunk 6]
```diff
@@ -245,7 +245,7 @@ def test_pad_with_tuple_of_pad_values(self):
 
     def test_pad_with_non_constant_padding_modes(self):
         """Unit tests for edge, reflect, symmetric padding"""
-        img = torch.zeros(3, 27, 27)
+        img = torch.zeros(3, 27, 27).byte()
```
Comment (fmassa) [inline test/test_transforms.py]: Those are merge conflicts. Can you please revert these changes?

File: test/test_utils.py [hunk 7]
```diff
@@ -18,14 +18,6 @@ def test_make_grid_not_inplace(self):
         utils.make_grid(t, normalize=True, scale_each=True)
         assert torch.equal(t, t_clone), 'make_grid modified tensor in-place'
 
-    def test_make_grid_raises_with_variable(self):
```
Comment (fmassa) [inline test/test_utils.py]: Merge conflict?

File: torchvision/datasets/hpatches.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Thanks @fmassa for pointing it out. 
From the docs 
> By default, each worker will have its PyTorch seed set to base_seed + worker_id, where base_seed is a long generated by main process using its RNG. However, seeds for other libraies may be duplicated upon initializing workers (w.g., NumPy), causing each worker to return identical random numbers. (See My data loader workers return identical random numbers section in FAQ.) You may use torch.initial_seed() to access the PyTorch seed for each worker in worker_init_fn, and use it to set other seeds before data loading.

So isn't it better to use 

```python
def worker_init_fn(worker):
    np.random.seed(torch.initial_seed())
```

[the remark this answers is not in the document]
Comment (vbalnt) [reply torchvision/datasets/hpatches.py]: Also what happens if someone forgets to call this? As I understand this the default behaviour (i..e without passing this) would be problematic right?
